File: src/main/scala/com/snowflake/snowpark/types/StructType.scala
package com.snowflake.snowpark.types

import com.snowflake.snowpark.internal.analyzer.Attribute
import com.snowflake.snowpark.internal.analyzer

/**
 * StructType data type, represents table schema.
 * @since 0.1.0
 */
object StructType {

  private[snowpark] def fromAttributes(attrs: Seq[Attribute]): StructType =
    StructType(attrs.map(a => StructField(a.name, a.dataType, a.nullable)))

  /**
   * Clones the given [[StructType]] object.
   * @since 0.1.0
   */
  def apply(other: StructType): StructType = StructType(other.fields)

  /**
   * Creates a [[StructType]] object based on the given Seq of [[StructField]]
   * @since 0.1.0
   */
  def apply(fields: Seq[StructField]): StructType = StructType(fields.toArray)

  /**
   * Creates a [[StructType]] object based on the given [[StructField]]
   * @since 0.7.0
   */
  def apply(field: StructField, remaining: StructField*): StructType = apply(field +: remaining)
}

/**
 * StructType data type, represents table schema.
 * @constructor Creates a new [[StructType]] object based on the given array of [[StructField]].
 * @since 0.1.0
 */
case class StructType(fields: Array[StructField] = Array())
    extends DataType
    with Seq[StructField] {

  /**
   * Returns the total number of [[StructField]]
   * @since 0.1.0
   */
  override def length: Int = fields.length

  /**
   * Converts this object to Iterator.
   * @since 0.1.0
   */
  override def iterator: Iterator[StructField] = fields.toIterator

  /**
   * Returns the corresponding [[StructField]] of the given index.
   * @since 0.1.0
   */
  override def apply(idx: Int): StructField = fields(idx)

  /**
   * Returns a String values to represent this object info.
   * @since 0.1.0
   */
  override def toString: String =
    s"StructType[${fields.map(_.toString).mkString(", ")}]"

  /**
   * Appends a new [[StructField]] to the end of this object.
   * @since 0.1.0
   */
  def add(field: StructField): StructType = StructType(fields :+ field)

  /**
   * Appends a new [[StructField]] to the end of this object.
   * @since 0.1.0
   */
  def add(name: String, dataType: DataType, nullable: Boolean = true): StructType =
    add(StructField(name, dataType, nullable))

  /**
   * (Scala API Only) Returns a Seq of the name of [[StructField]].
   * @since 0.1.0
   */
  def names: Seq[String] = fields.map(_.name)

  /**
   * Returns the corresponding [[StructField]] object of the given name.
   * @since 0.1.0
   */
  def nameToField(name: String): Option[StructField] =
    fields.find(_.columnIdentifier.quotedName == analyzer.quoteName(name))

  /**
   * Return the corresponding [[StructField]] object of the given name.
   * @since 0.1.0
   */
  def apply(name: String): StructField =
    nameToField(name).getOrElse(
      throw new IllegalArgumentException(s"$name does not exits. Names: ${names.mkString(", ")}"))

  protected[snowpark] def toAttributes: Seq[Attribute] = {
    /*
     * When user provided schema is used in a SnowflakePlan, we have to
     * to invoke quoteName on the column names to align with the
     * case-insensitive(unless quoted) identifier naming that snowflake follows.
     */
    map(f => Attribute(f.columnIdentifier.quotedName, f.dataType, f.nullable))
  }

  /**
   * Prints the StructType content in a tree structure diagram.
   * @since 0.9.0
   */
  def printTreeString(): Unit =
    // scalastyle:off
    println(treeString(0))
  // scalastyle:on

  private[snowpark] def treeString(layer: Int): String =
    (if (layer == 0) "root\n" else "") + fields.map(_.treeString(layer)).mkString

}

/**
 * Constructors and Util functions of [[StructField]]
 * @since 0.1.0
 */
object StructField {

  /**
   * Creates a [[StructField]]
   *
   * @since 0.1.0
   */
  def apply(name: String, dataType: DataType, nullable: Boolean): StructField =
    StructField(ColumnIdentifier(name), dataType, nullable)

  /**
   * Creates a [[StructField]]
   *
   * @since 0.1.0
   */
  def apply(name: String, dataType: DataType): StructField =
    StructField(ColumnIdentifier(name), dataType)
}

/**
 * Represents the content of [[StructType]].
 * @since 0.1.0
 */
case class StructField(
    columnIdentifier: ColumnIdentifier,
    dataType: DataType,
    nullable: Boolean = true) {

  /**
   * Returns the column name.
   * @since 0.1.0
   */
  val name: String = columnIdentifier.name

  /**
   * Returns a String values to represent this object info.
   * @since 0.1.0
   */
  override def toString: String = s"StructField($name, $dataType, Nullable = $nullable)"

  private[types] def treeString(layer: Int): String = {
    val prepended: String = (1 to (1 + 2 * layer)).map(x => " ").mkString + "|--"
    val body: String = s"$name: ${dataType.typeName} (nullable = $nullable)\n" +
      (dataType match {
        case st: StructType => st.treeString(layer + 1)
        case _ => ""
      })

    prepended + body
  }
}

/**
 *  Constructors and Util functions of ColumnIdentifier
 *  @since 0.1.0
 */
object ColumnIdentifier {

  /**
   * Creates a [[ColumnIdentifier]] object for the giving column name.
   * Identifier Requirement can be found from
   * https://docs.snowflake.com/en/sql-reference/identifiers-syntax.html
   * @since 0.1.0
   */
  def apply(name: String): ColumnIdentifier =
    new ColumnIdentifier(analyzer.quoteName(name))

  /**
   * Removes the unnecessary quotes from name
   *
   * Remove quotes if name starts with _A-Z and only contains _0-9A-Z$, or
   * starts with $ and follows by digits
   * @since 0.1.0
   */
  private def stripUnnecessaryQuotes(str: String): String = {
    val removeQuote = "^\"(([_A-Z]+[_A-Z0-9$]*)|(\\$\\d+))\"$".r
    str match {
      case removeQuote(n, _, _) => n
      case n => n
    }
  }
}

/**
 * Represents Column Identifier
 * @since 0.1.0
 */
class ColumnIdentifier private (normalizedName: String) {

  /**
   * Returns the name of column.
   * Name format:
   * 1. if the name quoted.
   *   a. starts with _A-Z and follows by _A-Z0-9$: remove quotes
   *   b. starts with $ and follows by digits: remove quotes
   *   c. otherwise, do nothing
   * 2. if not quoted.
   *   a. starts with _a-zA-Z and follows by _a-zA-Z0-9$, upper case all letters.
   *   b. starts with $ and follows by digits, do nothing
   *   c. otherwise, quote name
   *
   * More details can be found from
   * https://docs.snowflake.com/en/sql-reference/identifiers-syntax.html
   * @since 0.1.0
   */
  val name: String = ColumnIdentifier.stripUnnecessaryQuotes(normalizedName)

  /**
   * Returns the quoted name of this column
   * Name Format:
   * 1. if quoted, do nothing
   * 2. if not quoted.
   *   a. starts with _a-zA-Z and follows by _a-zA-Z0-9$, upper case all letters
   *     and then quote
   *   b. otherwise, quote name
   *
   * It is same as [[name]], but quotes always added.
   * It is always safe to do String comparisons between quoted column names
   * @since 0.1.0
   */
  def quotedName: String = normalizedName

  /**
   * Returns a copy of this [[ColumnIdentifier]].
   * @since 0.1.0
   */
  override def clone(): AnyRef = new ColumnIdentifier(normalizedName)

  /**
   * Returns the hashCode of this [[ColumnIdentifier]].
   * @since 0.1.0
   */
  override def hashCode(): Int = normalizedName.hashCode

  /**
   * Compares this [[ColumnIdentifier]] with the giving one, returns true if these
   * two are equivalent, otherwise, returns false.
   * @since 0.1.0
   */
  override def equals(obj: Any): Boolean =
    obj match {
      case other: ColumnIdentifier => normalizedName == other.quotedName
      case _ => false
    }

  /**
   * Returns the column name. Alias of [[name]]
   * @since 0.1.0
   */
  override def toString: String = name
}

File: src/main/java/com/snowflake/snowpark_java/types/StructType.java
package com.snowflake.snowpark_java.types;

import com.snowflake.snowpark.internal.JavaUtils;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Optional;

/**
 * StructType data type, represents table schema.
 *
 * @since 0.9.0
 */
public class StructType extends DataType implements Iterable<StructField> {
  private final com.snowflake.snowpark.types.StructType scalaStructType;

  StructType(com.snowflake.snowpark.types.StructType structType) {
    this.scalaStructType = structType;
  }

  /**
   * Clones the given StructType object.
   *
   * @param other A StructType object
   * @since 0.9.0
   */
  public StructType(StructType other) {
    this(other.scalaStructType);
  }

  /**
   * Creates a StructType object based on the given Array of StructField.
   *
   * @param fields An Array of StructField
   * @since 0.9.0
   */
  public StructType(StructField[] fields) {
    this(new com.snowflake.snowpark.types.StructType(toScalaFieldsArray(fields)));
  }

  /**
   * Creates a StructType object based on the given StructField
   *
   * @param fields A list of StructFields
   * @since 0.9.0
   * @return A new StructType object
   */
  public static StructType create(StructField... fields) {
    return new StructType(fields);
  }

  private static com.snowflake.snowpark.types.StructField[] toScalaFieldsArray(
      StructField[] fields) {
    com.snowflake.snowpark.types.StructField[] result =
        new com.snowflake.snowpark.types.StructField[fields.length];
    for (int i = 0; i < result.length; i++) {
      result[i] = fields[i].toScalaStructField();
    }
    return result;
  }

  /**
   * Return the index of the specified field.
   *
   * @param fieldName the name of the field.
   * @return the index of the field with the specified name.
   * @throws IllegalArgumentException if the given field name does not exist in the schema.
   * @since 1.15.0
   */
  public int fieldIndex(String fieldName) {
    return this.scalaStructType.fieldIndex(fieldName);
  }

  /**
   * Retrieves the names of all {@link com.snowflake.snowpark_java.types.StructField} in this {@link
   * com.snowflake.snowpark_java.types.StructType}.
   *
   * <p>Example:
   *
   * <pre>{@code
   * StructType schema = new StructType(new StructField[] {
   *   new StructField("c1", DataTypes.IntegerType),
   *   new StructField("c2", DataTypes.StringType),
   * });
   * schema.names();
   * // res: String[] = {"C1", "C2"}
   * }</pre>
   *
   * @return an array representing the names of the fields in this StructType.
   * @since 0.9.0
   */
  public String[] names() {
    return JavaUtils.seqToJavaStringArray(this.scalaStructType.names());
  }

  /**
   * Retrieves the names of all {@link com.snowflake.snowpark_java.types.StructField} in this {@link
   * com.snowflake.snowpark_java.types.StructType}. This is an alias of {@link #names}.
   *
   * <p>Example:
   *
   * <pre>{@code
   * StructType schema = new StructType(new StructField[] {
   *   new StructField("c1", DataTypes.IntegerType),
   *   new StructField("c2", DataTypes.StringType),
   * });
   * schema.fieldNames();
   * // res: String[] = {"C1", "C2"}
   * }</pre>
   *
   * @return an array representing the names of the fields in this StructType.
   * @since 1.17.0
   */
  public String[] fieldNames() {
    return this.names();
  }

  /**
   * Counts the number of StructFields in this StructType object.
   *
   * @return An int number
   * @since 0.9.0
   */
  public int size() {
    return this.scalaStructType.size();
  }

  /**
   * Creates an Iterator of StructFields.
   *
   * @return An Iterator of StructField
   * @since 0.9.0
   */
  @Override
  public Iterator<StructField> iterator() {
    return new Iterator<StructField>() {
      private int index = 0;

      @Override
      public boolean hasNext() {
        return scalaStructType.size() > index;
      }

      @Override
      public StructField next() {
        if (!hasNext()) {
          throw new NoSuchElementException(
              "Index: " + index + " iterator size: " + scalaStructType.size());
        }
        return new StructField(scalaStructType.apply(index++));
      }
    };
  }

  /**
   * Retrieves the StructField object from the given index.
   *
   * @param index An int number representing the index of StructField
   * @return The StructField object at the given index.
   * @since 0.9.0
   */
  public StructField get(int index) {
    return new StructField(this.scalaStructType.apply(index));
  }

  /**
   * Generates a String value to represent this StructType.
   *
   * @return A String value in the format of "StructType[StructField('name', 'data type',
   *     'nullable')...]"
   * @since 0.9.0
   */
  @Override
  public String toString() {
    return this.scalaStructType.toString();
  }

  /**
   * Creates new StructType by appending the given StructField to the end of this StructType. This
   * function doesn't modify this StructType object, but create and return a new one.
   *
   * @param field The StructField object being appended.
   * @return A new StructType object
   * @since 0.9.0
   */
  public StructType add(StructField field) {
    return new StructType(this.scalaStructType.add(field.toScalaStructField()));
  }

  /**
   * Creates new StructType by appending a new StructField with the given info to the end of this
   * StructType. This function doesn't modify this StructType object, but create or return a new
   * one.
   *
   * @param name The name of the StructField object being appended.
   * @param dataType The data type of the StructField object being appended.
   * @param nullable Whether the new StructField is nullable or not
   * @return A new StructType object
   * @since 0.9.0
   */
  public StructType add(String name, DataType dataType, boolean nullable) {
    return add(new StructField(name, dataType, nullable));
  }

  /**
   * Creates new StructType by appending a new StructField with the given info to the end of this
   * StructType. The new StructField is nullable. This function doesn't modify this StructType
   * object, but create and return a new one.
   *
   * @param name The name of the StructField object being appended.
   * @param dataType The data type of the StructField object being appended.
   * @return A new StructType object
   * @since 0.9.0
   */
  public StructType add(String name, DataType dataType) {
    return add(name, dataType, true);
  }

  /**
   * Retrieves the corresponding StructField object of the given name.
   *
   * @param name The name of StructField
   * @return An Optional StructField object.
   * @since 0.9.0
   */
  public Optional<StructField> nameToField(String name) {
    scala.Option<com.snowflake.snowpark.types.StructField> fields =
        this.scalaStructType.nameToField(name);
    return fields.isEmpty() ? Optional.empty() : Optional.of(new StructField(fields.get()));
  }

  /**
   * Verifies if a StructType equals to this one.
   *
   * @param other A StructType object
   * @return true if these data types are equivalent, false for otherwise.
   * @since 0.9.0
   */
  @Override
  public boolean equals(Object other) {
    if (other instanceof StructType) {
      return this.scalaStructType.equals(((StructType) other).scalaStructType);
    }
    return false;
  }

  /**
   * Calculates the hash code of this StructType Object.
   *
   * @return An int number representing the hash code value
   * @since 0.9.0
   */
  @Override
  public int hashCode() {
    return this.scalaStructType.hashCode();
  }

  /**
   * Prints the schema StructType content in a tree structure diagram.
   *
   * @since 0.9.0
   */
  public void printTreeString() {
    scalaStructType.printTreeString();
  }

  com.snowflake.snowpark.types.StructType getScalaStructType() {
    return scalaStructType;
  }
}
